### PR TITLE
Use bytes in arrow-flight

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change
 

Prost supports zero-copy reading of binary payloads, internally slicing the original `Bytes` comprising the HTTP frame read by hyper. This PR tweaks prost-build to generate structs with `Bytes` instead of `Vec<u8>` which should cut down on the amount of data copying during decode.

TBC there will still be copies resulting from:

* `prost_types::Any` uses `Vec` not `Bytes` (we could work around this in a subsequent PR)
* Conversion to `Buffer` involves a memcpy

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-flight/examples/flight_sql_server.rs [hunk 1]
```diff
@@ -23,18 +23,14 @@ use arrow_flight::{
     Location, SchemaAsIpc, Ticket,
 };
 use futures::{stream, Stream};
+use prost::Message;
```
Comment (tustvold) [inline arrow-flight/examples/flight_sql_server.rs:26]: Most of the changes in this file are actually clippy lints that weren't being picked up by CI (which I have also fixed)

File: arrow-flight/build.rs [hunk 2]
```diff
@@ -94,3 +94,9 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
     // As the proto file is checked in, the build should not fail if the file is not found
     Ok(())
 }
+
+fn prost_config() -> prost_build::Config {
```
Comment (tustvold) [inline arrow-flight/build.rs:98]: Unfortunately `prost_build::Config` is not `Clone` so this was my workaround

File: arrow-flight/build.rs [hunk 3]
```diff
@@ -94,3 +94,9 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+    let mut config = prost_build::Config::new();
+    config.bytes([".arrow"]);
```
Comment (alamb) [inline arrow-flight/build.rs:100]: Maybe we should put a note / link that this uses the `Bytes` API to avoid copies

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @tustvold 

I recommend we wait for a few days for more feedback on this API prior to merging

We could also perhaps add some documentation changes / explanation in the PR or code to help migration.

cc @avantgardnerio I suspect this will require some changes downstream in Ballista

**Review by Dandandan — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by tustvold** [comment]
I plan to merge this this afternoon unless there are any objections

**Pull request merged**